### PR TITLE
Add JobApplicationCard component

### PR DESCRIPTION
**Pull request by rayepps**
**Release Type:** *Non-Breaking Feature*

Fixes https://x-team-internal.atlassian.net/browse/XP-3373

## Description

Add JobApplicationCard component.

## Checklist

- [x] set yourself as an assignee
- [x] set appropriate labels for a PR (`In Review` or `In Progress` depending on its status)
- [x] move respective JIRA issue to the `IN REVIEW` column
- [x] make sure your code lints (`npm run lint`)
- [x] Flow typechecks passed (`npm run flow`)
- [x] Snapshots tests passed (`npm run jest`)
- [x] check cleanup tasks (https://github.com/x-team/xp/labels/cleanup) and take a suitable small one (if exists) in a related area of the current changes
- [x] component's documentation (`.stories.js` file) is changed or added accordingly to reflect any new or updated use cases or variants usage
- [x] **manually tested the app** by running it in several different browsers (Firefox, Chrome, Opera, Safari, MS IE/Edge, etc.) and checked nothing is broken and operates as expected!

## Steps to Test or Reproduce

Check out the storybook site and look at `JobApplicationCard` component stories. Compare to mockup images on the ticket
[XP-3373](https://x-team-internal.atlassian.net/browse/XP-3373)
and on design
https://www.figma.com/file/ilfpDrEx9nKf7BXEQffGqu/XP-Jobs

## Screenshots

Screen Shot 2019-10-21 at 8 41 56 PM
Screen Shot 2019-10-21 at 8 42 05 PM

### CODE REVIEW

File: src/components/ui/JobApplicationCard.stories.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  ))
+
+storiesOf('UI Components|JobApplicationCard/Debug', module)
+  .add('missing date', () => (
+    <JobApplicationCard
+      hasApplied
+      onWithdraw={action('Withdrawing from job...')}
+    />
+  ))
+  .add('missing props (does component explode?)', () => (
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.stories.js]: Let's shorten this to just `missing props` as we agreed on a new task you created couple days ago.

File: src/components/ui/JobApplicationCard.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  applicationDate,
+  onApply,
+  onWithdraw
+}: Props) => {
+  const date = () => {
+    const month = getMonthName(applicationDate)
+    const day = getDayWithOrdinal(applicationDate)
+    return `${month} ${day}`
+  }
+  const applyCard = () => (
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.js]: Usually, when a method name starts with a verb this means that action is gonna be performed, so that's what expected to be the outcome when skimming through the code. In your case, as I can see, we just render stuff, so it's suggested we rename methods to be in line with `renderSmth`.

File: src/components/ui/JobApplicationCard.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  const withdrawCard = () => (
+    <Fragment>
+      <Button wide color='monochrome' onClick={onWithdraw}>Widthdraw Application</Button>
+      <p className={cx.paragraph}>
+        You've already applied for this position{ applicationDate && ` on ${date()}` }.
+        <b> Check your email for next steps.</b>
+      </p>
+    </Fragment>
+  )
+  return (
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.js]: Could you please separate code blocks with empty lines to make code easier to read? Adding empty line before `render` is usually a good option to have.

File: src/components/ui/JobApplicationCard.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    <div className={cx.card}>
+      { hasApplied ? withdrawCard() : applyCard() }
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.js]: Optional change request: to keep code consistent, could we remove spaces inside curly braces? We usually do that when destructuring. When the main code logic goes, it can go without spaces.

File: src/components/ui/JobApplicationCard.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    return `${month} ${day}`
+  }
+  const applyCard = () => (
+    <Button wide onClick={onApply}>Apply For This Position</Button>
+  )
+  const withdrawCard = () => (
+    <Fragment>
+      <Button wide color='monochrome' onClick={onWithdraw}>Widthdraw Application</Button>
+      <p className={cx.paragraph}>
+        You've already applied for this position{ applicationDate && ` on ${date()}` }.
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.js]: seems like you have a syntactical error here? The code even seems to be highlighted wrong by Github. I guess you meant to use `${}` inside template literal or smth. Did it work for you when you tested?

File: src/components/ui/JobApplicationCard.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+const cmz = require('cmz')
+
+const cx = {
+  card: cmz(`
+    max-width: 220px
+    padding: 20px
+    display: inline-block
+    background: ${theme.baseBrighter}
+    border: 1px solid ${theme.lineSilver5}
+    box-shadow: 4px 4px 0px ${theme.baseBrightSilver}
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.js]: For unitless values units can be skipped, so that: `0px` is the same as `0`. Feel free to update the declaration above.

File: src/components/ui/JobApplicationCard.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
+// @flow
+
+import React, { Fragment } from 'react'
+import theme from '../../styles/theme'
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.js]: Can I suggest to split imports a bit to make it consistent with the rest of files? External dependencies first, then any local ones, then styles, types, etc:

```js
import React, { Fragment } from 'react'

import Button from './Button'

import theme from '../../styles/theme'
import typo from '../../styles/typo'
```

Just a suggestion to make code easier to read 😛

File: src/components/ui/JobApplicationCard.js (hunk 5 above)
Comment (rayepps) [reply src/components/ui/JobApplicationCard.js]: This is correct and yes it works

File: src/components/ui/JobApplicationCard.stories.js (hunk 1 above)
Comment (rayepps) [reply src/components/ui/JobApplicationCard.stories.js]: done in [19e7010](https://github.com/x-team/xp-ui/pull/462/commits/19e70104f3e987f3a1013587a86e0d3cbd08eeb3)

File: src/components/ui/JobApplicationCard.js (hunk 2 above)
Comment (rayepps) [reply src/components/ui/JobApplicationCard.js]: done in [19e7010](https://github.com/x-team/xp-ui/pull/462/commits/19e70104f3e987f3a1013587a86e0d3cbd08eeb3)

File: src/components/ui/JobApplicationCard.js (hunk 3 above)
Comment (rayepps) [reply src/components/ui/JobApplicationCard.js]: done in [19e7010](https://github.com/x-team/xp-ui/pull/462/commits/19e70104f3e987f3a1013587a86e0d3cbd08eeb3)

File: src/components/ui/JobApplicationCard.js (hunk 6 above)
Comment (rayepps) [reply src/components/ui/JobApplicationCard.js]: done in [19e7010](https://github.com/x-team/xp-ui/pull/462/commits/19e70104f3e987f3a1013587a86e0d3cbd08eeb3)

File: src/components/ui/JobApplicationCard.js (hunk 7 above)
Comment (rayepps) [reply src/components/ui/JobApplicationCard.js]: done in [19e7010](https://github.com/x-team/xp-ui/pull/462/commits/19e70104f3e987f3a1013587a86e0d3cbd08eeb3)

**Comment by rayepps** [comment]
@nicksp @bernardodiasc made PR changes and committed build files - ready for review

**Comment by rayepps** [comment]
> rayepps You told you added new component to the to the registration bundle, where can I find that?
I don't know what are you talking about. what registration bundle?

**Comment by nicksp** [comment]
> what registration bundle?

The way we expose components. Please check how other components are exposed for public use.

**Comment by rayepps** [comment]
@nicksp I added new component to index and registration

File: src/components/ui/JobApplicationCard.stories.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
+// @flow
+
+import React from 'react'
+import { storiesOf } from '@storybook/react'
+import { action } from '@storybook/addon-actions'
+
+import JobApplicationCard from './JobApplicationCard'
+
+storiesOf('UI Components|JobApplicationCard', module)
+  .add('basic usage', () => (
```
Comment (bernardodiasc) [inline src/components/ui/JobApplicationCard.stories.js]: This is not a requirement, just a suggestion. Because thinking on this story to be actually useful for non-dev folks, we should be able to demonstrate a bit further. Sometimes using a single state of the component is not enough, so we have an special component to simulate a contextual state, this component is meant to be used on on stories: https://github.com/x-team/xp-ui/blob/develop/src/utils/State.js

You can check how other components uses it, for instance the StatusFilter: https://github.com/x-team/xp-ui/blob/develop/src/components/ui/StatusFilter.stories.js#L8-L24, see it live: https://xpui-develop.netlify.com/?path=/story/ui-components-statusfilter--basic-usage

For this case it would be nice to see the `hasApplied` switching when clicking in the button, and optionally if want to make a even better showcase, add a timer to show the `isSaving` state in between.

**Comment by rayepps** [comment]
@bernardodiasc I updated the component to be 100% wide 👍

File: src/components/ui/JobApplicationCard.stories.js (hunk 8 above)
Comment (nicksp) [reply src/components/ui/JobApplicationCard.stories.js]: Well, it's a storybook, so having some interactive experience seems to be beneficial for the design/product team. It seems to make no harm to the _actual_ components, more of a presentational state that gives a better clue of expectations. I'd go for this suggestion 🚀

Comment (bernardodiasc) [reply src/components/ui/JobApplicationCard.stories.js]: Don't need to do this move. Maybe it's not clear the reason we do this. The point is the many times we don't use local state on UI components, leaving them to be more static. The state in many times is a responsibility of the consumer application, that uses some data management like Redux. Having a story demonstrating how an outer-state would interact with the UI component has proven many times to be useful for developer to understand how the data management should be construct on it's own platform. Also it's useful to demonstrate the component working and helpful for debug state updates. Imagine the Modal showcase without the ability to see closing, it hides an important part of the presentation component functionality. Speaking for myself, I construct stories to help me develop the components, the certainly helps me to have a simulation of an app state to interact with the UI component.

But as I said, this is not a blocker. If you don't feel good doing this we can leave it out.

**Comment by bernardodiasc** [comment]
@rayepps when you created the new story for this component https://deploy-preview-462--xpui-develop.netlify.com/?path=/story/ui-components-jobapplicationcard-use-cases--contained-in-column-like-it-is-in-the-job-page-layout have you had a change to see that it looks wrong?

Screen Shot 2019-10-25 at 07 20 23

I mean, you did good move, that exactly the point of developing in storybook, we can see the issues before delivering. In this the `white-space: normal` would make it better.

Screen Shot 2019-10-25 at 07 26 22

File: src/components/ui/JobApplicationCard.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    return `${month} ${day}`
+  }
+
+  const renderApplyCard = () => (
+    <Button wide disabled={isSaving} onClick={onApply}>Apply For This Position</Button>
+  )
+
+  const renderWithdrawCard = () => (
+    <Fragment>
+      <Button wide color='monochrome' disabled={isSaving} onClick={onWithdraw}>Widthdraw Application</Button>
```
Comment (nicksp) [inline src/components/ui/JobApplicationCard.js]: There's a typo
```suggestion
      <Button wide color='monochrome' disabled={isSaving} onClick={onWithdraw}>Withdraw Application</Button>
```

File: src/registration.js [hunk 10]
```diff
@@ -19,6 +19,7 @@ import HorizontalRuler from './components/ui/HorizontalRuler'
 import InputField from './components/forms/InputField'
 import InputGroup from './components/forms/InputGroup'
 import IssueCollectorButton from './components/ui/IssueCollectorButton'
+import JobApplicationCard from './components/ui/JobApplicationCard'
```
Comment (nicksp) [inline src/registration.js]: Let's please also add this to `src/index.js` which is our `all` bundle: https://github.com/x-team/xp-ui/blob/fc9772271e5cc0a3711065457b9c9c3d9dd95751/webpack.standalone.js#L7

Even if it might not be needed, we should reflect it's purpose which is a bundle to hold all the public components.

Comment (bernardodiasc) [reply src/registration.js]: Interesting to know about that, I wasn't aware. So **all**  the components be exported in the `index.js`?

Comment (nicksp) [reply src/registration.js]: Not literally **all** components, only those that are meant to be imported directly, public ones if you wish. Pretty much everything we do have in `registration` should probably also go inside `index.js`. The use case might be just deciding to use any of the public components from inside another app integration using the `all` bundle, but we suddenly realize it's not there. I hope I made the point clear :) Again, those components that we don't use **directly** but rather use inside UI components themselves, shouldn/t be exported from anywhere.
